lib/acts_as_ar_query.rb

```diff
@@ -0,0 +1,205 @@
+  # - [X] find
+  # - [ ] find_by
+  # - [X] first
+  # - [X] last
+  # - [X] size
+  # - [X] take
+
+  # methods that enhance / chain the query
+  # list is from ActiveRecord relation interface
+
+  # - [ ] bind
+  # - [ ] create_with
+  # - [ ] distinct
+  # - [ ] eager_load
+  # - [X] except ? - is this not defined in the interface?
+  # - [ ] extending
+  # - [ ] from
+  # - [X] group
+  # - [ ] ~~having~~ - NO
+  # - [X] includes (partial)
+  # - [ ] joins
+  # - [X] limit
+  # - [ ] lock
+  # - [.] none
+  # - [X] offset
+  # - [X] order (partial)
+  # - [ ] preload
+  # - [ ] readonly
+  # - [X] references (partial)
+  # - [X] reorder
+  # - [ ] reverse_order
+  # - [X] select (partial)
+  # - [X] unscope
+  # - [ ] uniq
+  # - [X] where (partial)
+  # - [ ] where.not
+
+  # private api
+  # the common pattern is to allow this method to be called frequently
+  # and add the values into the hash
+  # this is used by methods like includes, references, order, select and others
+  def append_hash_arg(*val)
+    symbol = __callee__
+    val = val.flatten
+    if val.first.kind_of?(Hash)
+      raise ArgumentError, "Need to support #{symbol}(#{val.class.name})"
+    end
+    dup.tap do |r|
+      r.options[symbol] = r.options[symbol] ? (r.options[symbol] + val) : val
+    end
+  end
+
+  def assign_arg(val)
+    dup.tap do |r|
+      r.options[__callee__] = val
+    end
+  end
+
+  # public api
+
+  def initialize(model, opts = {})
+    @klass   = model
+    @options = opts || {}
+  end
+
+  # @param val [Hash]
+  # TODO [Array, Hash, String]
+  def where(*val)
+    val = val.flatten
+    val = val.first if val.size == 1 && val[0].kind_of?(Hash)
+    dup.tap do |r|
+      old_where = r.options[:where]
+      if val.empty?
+        # nop
+      elsif old_where.blank?
+        r.options[:where] = val
+      elsif old_where.kind_of?(Hash) && val.kind_of?(Hash)
+        val.each_pair do |key, value|
+          old_where[key] = if old_where[key]
+                             Array.wrap(old_where[key]) + Array.wrap(value)
+                           else
+                             value
+                           end
+        end
+      else
+        raise ArgumentError,
+              "Need to support #{__callee__}(#{val.class.name}) with existing #{old_where.class.name}"
+      end
+    end
+  end
+
+  # @param val [Array] for includes
+  # TODO: support hash
+  alias includes append_hash_arg
+
+  # @param val Array for includes
+  alias references append_hash_arg
+
+  # @param [Integer] val
+  alias limit assign_arg
+
+  # @param [String, Symbol, Arel] val
+  # TODO: support Hash
+  # TODO: support mixing?
+  alias order append_hash_arg
+  alias group append_hash_arg
+
+  def reorder(*val)
+    val = val.flatten
+    if val.first.kind_of?(Hash)
+      raise ArgumentError, "Need to support #{__callee__}(#{val.class.name})"
+    end
+
+    dup.tap do |r|
+      r.options[:order] = val
+    end
+  end
+
+  def except(*val)
+    val = val.flatten
```

Comment:
`val.flatten!` ?   Since the Array is constructed on the fly because of the splat operator, then there is no harm and it won't modify the caller's object.